### DISCUSSION TITLE
How to set default browser if Url not match?

### DISCUSSION THREAD

**Discussion by yoda65** [nref/BrowseRouter#17] [Q&A]
Hi,
if no URL match is found in [Urls] section, the URL does not launch. 
(log message : 01/10/2023 20:56:08 BrowseRouter: Unable to find a browser matching "my Url")
How to set the default browser to use  if no URL (chrome for example) match is found?
The way I found is to set '*=chrome' at the end of the [URLs] section. This seems to work.
Is there another method?

I am looking at your software to complete a lack of keepassXC software (opensource password manager) which does not allow  to choose a specific browser for an entry.

Regards

**Comment by nref** [comment]
Hey @yoda65 , thanks for reaching out.

I use KeepassXC every day, in combination with BrowseRouter. It's a powerful combination!

If there is no match for a URL, the app falls back to the catch-all URL. If there is no catch-all URL in `config.ini`, the app falls back to the *first* listed browser in `[browsers]`. When you add e.g. `* = chrome` you override the catch-all URL.

Example:

```ini
[browsers]
ff = C:\Program Files\Mozilla Firefox\firefox.exe
edge = C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe

[urls]
youtube.com = edge
;* = ff ; implied
```

Tested just now:

```powershell
PS C:\Program Files\BrowseRouter> .\BrowseRouter.exe youtube.com # opens in Edge
PS C:\Program Files\BrowseRouter> .\BrowseRouter.exe github.com # opens in FireFox
```

Log file:
```log
10/3/2023 10:12:47 BrowseRouter: Attempting to launch "youtube.com" for "Administrator: PowerShell"
10/3/2023 10:12:47 BrowseRouter: Found URL preference "youtube.com" => "edge" ("C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe")
10/3/2023 10:12:47 BrowseRouter: Launching C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe with args " https://youtube.com"

10/3/2023 10:12:56 BrowseRouter: Attempting to launch "github.com" for "Administrator: PowerShell"
10/3/2023 10:12:56 BrowseRouter: Found URL preference "*" => "ff" ("C:\Program Files\Mozilla Firefox\firefox.exe")
10/3/2023 10:12:56 BrowseRouter: Launching C:\Program Files\Mozilla Firefox\firefox.exe with args " https://github.com"
```
If the problem persists, can you provide a minimal config.ini? Thank you!

**Reply by yoda65** [reply]
Hi @slater1 ,

Thanks a lot.
I tested your config file sample:
The exe downloaded from the repository works fine.
but i wanted to rebuilt the project (As i do very ofen with opensource)
So I downloaded sources I rebuilt the project in the last Visual Studio 2022 free version. (with publish menu)
The new exe encounters the problem described above.
I publish without any change.
There are no built or publish error.
I use a fresh install VS 2022 in VM virtual Box Win10
The size of the two Exe (downloaded from the repo vs newer one) are not exactly the same.

I'm not a Visual Studio expert. I suppose there is a setup or VS version problem or it's not the good way to do...

Regards

**Comment by nref** [comment]
Hey @yoda65 ,

The current release is older than the tip of master. That might explain the binary mismatch. But I make no effort to prove the authenticity of builds. There's no CI currently and no signing. If you need zero-trust, best to build BrowseRouter yourself, as you're doing. Publishing with VS 2022 on Win10 is one supported way. 

It's quite possible there is a regression on the master branch. I will have a look ASAP.
